### DISCUSSION TITLE
BelongsTo field resets after changing value when depends on itself

### DISCUSSION THREAD

**Discussion by kusab85** [laravel/nova-issues#5190] [Q&A] [answered]
- Laravel Version: 9.45.1 
- Nova Version:  4.19.5 (Silver Surfer)
- PHP Version: 8.0.11
- Database Driver & Version: SQL lite
- Operating System and Version: Windows 10 22H2
- Browser type and version: FF, Chrome
- Reproduction Repository: https://github.com/kusab85/TestApp/tree/nova-BelongsTo-resets-with-dependsOn

### Description:
Value of BelongsTo field resets to its initial value when dependsOn itself.

```php 
/**
 * App/Nova/Post.php 
 */

    public function fields(NovaRequest $request)
    {
        return [
            // [...]
            BelongsTo::make('User')
                ->dependsOn(
                    ['user'],
                    function (BelongsTo $field, NovaRequest $request, FormData $formData) {
                        $field->help(
                            '$formData->user = '.var_export($formData->user, true).' '.
                            '$field->value = '.var_export($field->value, true)
                        );
                    }),
           // [...]
        ];
    }
```
**1. Initial state of form field**
![obraz](https://user-images.githubusercontent.com/53086618/209807918-669abd1d-f362-44fb-a891-bcbb86489660.png)

**2. Selecting a new user**
![obraz](https://user-images.githubusercontent.com/53086618/209808224-2d1fd953-d2d2-4d86-9d45-aab3217ba44f.png)

**3. Form field state after new user selection**
![obraz](https://user-images.githubusercontent.com/53086618/209808502-8125d6c0-23a2-4e4d-8b55-19645f021639.png)

### Detailed steps to reproduce the issue on a fresh Nova installation:

Check reproduction repository.

**Comment by crynobone** [comment, marked as answer]
Yes, this is the expected behavior to avoid the initial default being populated and only consider `$field->value` assigned via `dependsOn()`

**Reply by kusab85** [reply]
I don't assign nor want to assign any value. I just want to set fields help text according to value change made manually by a user. 
Just look at the example above. I select "Chadrick Mante" (whose ID is 7) but fields display value is set back to "Mossie Mayer".

**Reply by crynobone** [reply]
Sorry but Dependent Field shouldn't listen to self, that just never considered.

**Reply by kusab85** [reply]
But it worked as I expected prior to 4.19.3 version. 
I think the fix _of issue where BelongsTo fields weren't broadcasting their changes_ changed fields behavior.

**Reply by crynobone** [reply]
It was never designed to work as you described.

**Reply by kusab85** [reply]
Ok, what can I say? I'll find a way around. ;-)

**Comment by ghost** [comment]
@kusab85 have you found a way around? :)

**Reply by ghost** [reply]
I have the exact same use case. And by the way listening to the field itself work fine with Text fields and I believe for Select fields as well.

**Reply by crynobone** [reply]
There were some significant changes on Dependent Fields for the last few patch releases so it could be possible. Still, it sure is not a behavior we cover when implementing this feature.

As long as depending on self and not doing any `$field->setValue()` you should be fine.

**Reply by ghost** [reply]
Thanks @crynobone! The issue I'm experiencing right now is that when selecting an option from the creation form, Nova shows the loading indicator (green line on top of the screen) and clears the select field. On the edit form it's even worse - it reloads the field indefinitely (the loading indicator is shown and bunch of ajax requests are sent in an infinite loop). The same thing happens when dependsOn() callback is entirely empty.

I'm on v4.25.1. 

```
public static function make(): BelongsTo
{
    return BelongsTo::make(__('Work Order Type'), 'workOrderType', WorkOrderTypeResource::class)
        ->dependsOn(
            'workOrderType',
            function (BelongsTo $field, NovaRequest $request, FormData $formData) {
                if (! $formData->workOrderType) {
                    return;
                }

                $workOrderType = WorkOrderType::query()->find($formData->workOrderType);
                $field->help($workOrderType->description);
            }
        );
}
```

**Reply by crynobone** [reply]
In that case it not possible at the moment.

**Comment by crynobone** [comment]
You should be able to depends on self in 4.26.0

**Reply by kusab85** [reply]
Hooray!

**Reply by kusab85** [reply]
But is this behavior meant to be permanent feature, @crynobone ? ;-)
